source/common/grpc/async_client_impl.cc

```diff
@@ -97,18 +97,18 @@ void AsyncStreamImpl::initialize(bool buffer_body_for_retry) {
 
 // TODO(htuch): match Google gRPC base64 encoding behavior for *-bin headers, see
 // https://github.com/envoyproxy/envoy/pull/2444#discussion_r163914459.
-void AsyncStreamImpl::onHeaders(Http::HeaderMapPtr&& headers, bool end_stream) {
+void AsyncStreamImpl::onHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) {
   const auto http_response_status = Http::Utility::getResponseStatus(*headers);
   const auto grpc_status = Common::getGrpcStatus(*headers);
-  callbacks_.onReceiveInitialMetadata(end_stream ? std::make_unique<Http::HeaderMapImpl>()
+  callbacks_.onReceiveInitialMetadata(end_stream ? std::make_unique<Http::ResponseHeaderMapImpl>()
                                                  : std::move(headers));
   if (http_response_status != enumToInt(Http::Code::OK)) {
     // https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md requires that
     // grpc-status be used if available.
     if (end_stream && grpc_status) {
-      // There is actually no use-after-move problem here,
-      // because it will only be executed when end_stream is equal to true.
-      onTrailers(std::move(headers)); // NOLINT(bugprone-use-after-move)
+      // Due to headers/trailers type differences we need to copy here. This is an uncommon case but
+      // we can potentially optimize in the future.
+      onTrailers(Http::createHeaderMap<Http::ResponseTrailerMapImpl>(*headers));
```

Comment:
can we extract the relevant part of `onTrailers` to `onGrpcTrailers(HeaderMapImplPtr)` and call them from both here and `onTrailers`?

Reply:
The issue here is that we have to change the physical type (turn headers into trailers) and I don't think extraction will help here? I was able to split out the logic in the transcoder to deal with a similar case. With all of that said, the next step is to move towards HeaderMap being an abstract base class, with relevant O(1) headers moved to where they are used, so I'm going to need to come back to this soon (gRPC headers/trailers confuse this a bit).

Reply:
OK sounds fair.